### PR TITLE
Add script for generating cert macros

### LINKED ISSUE #4914: Allow auto-generation of library/certs.c from tests/data_files
### Suggested enhancement

Annotates the hardcoded test certificates and keys in `library/certs.c` with a formal comment syntax describing their type and origin in `tests/data_files`, and adds a script `scripts/generate_certs.sh` which parses these comments and updates the hardcoded data. This aims to simplify the synchronization of `tests/data_files` with library/certs.c as well as the addition of new test data. This was suggested by @AndrzejKurek in #2260.

The work was started by @hanno-arm in #2596

### PR DESCRIPTION
**Pull request by yuhaoth**
## Description

fix #4914

`tests/src/certs.c` includes some ca certificates which come from files in `tests/data_files` . Those macros are update manually. 
That might cause out-of-sync issue. 

This script is to reduce the out-of-sync risk and reduce workload for `certificate expiry 2023q2` issues.

## Gatekeeper checklist

- [x] **changelog** not user visible
- [x] **backport** not required - nothing expiring in the 2.28 lifetime
- [x] **tests** not required

### CODE REVIEW

**Comment by yuhaoth** [comment]
I am not sure if this PR should be high priority. Please fix me if you do not think so.

**Comment by gilles-peskine-arm** [comment]
It would help since we're regenerating some of the content (the expiring certificates), but of course we can do the update manually this time. So I'd like to do this now unless we run out of time.

File: tests/data_files/test_certs.h.jinja2 [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        {% endfor %}
+
+/* END FILE */
+    {% endif %}
+    {% if mode == 'binary' %}
+/* This is generated from {{value}}. */
+/* BEGIN FILE binary macro {{name}} {{value}} */
+#define {{name}} {% raw -%} { {%- endraw %} {{ '\\' | put_to_column(position=80-11-name|length)}}
+    {% for line in value | read_as_c_array %}
+        {% if not loop.last %}
```
Comment (lpy4105) [inline tests/data_files/test_certs.h.jinja2:40]: Minor: need more indent.
Do you think we could remove the all the indents for `{% ... %}` blocks?

File: tests/data_files/Makefile [hunk 2]
```diff
@@ -1558,6 +1558,77 @@ pkcs7_data_cert_signeddata_sha256.der: pkcs7_data_cert_signed_sha256.der
 	dd if=pkcs7_data_cert_signed_sha256.der of=$@ skip=19 bs=1
 all_final += pkcs7_data_cert_signeddata_sha256.der
 
+################################################################
+#### Generate C format test certs header
+################################################################
+
+# FIXME: Makefile will update those input files which is not expected.
+#        Remove them from dependancy to workaround it.
+#        For timebeing, we will always update it when `make ../src/test_certs.h`
```
Comment (lpy4105) [inline tests/data_files/Makefile:1567]: Do we still need this workaround? And please be aware that string `PolarSSLTest` couldn't be a dependancy.

**Review by lpy4105 — COMMENTED** [review]
LGTM, except for some minor comments and the conflicts need to be resolved.

File: tests/data_files/test_certs.h.jinja2 (hunk 1 above)
Comment (yuhaoth) [reply tests/data_files/test_certs.h.jinja2:40]: I don't think so. That might be difficult to read.

Comment (lpy4105) [reply tests/data_files/test_certs.h.jinja2:40]:
```suggestion
        {% for line in value | read_as_c_array %}
            {% if not loop.last %}
```
Then I think we need 4 more space to correct the indentation here, as well as L42 and L44-45.

File: tests/data_files/Makefile [hunk 3]
```diff
@@ -1809,6 +1809,72 @@ pkcs7_data_cert_signeddata_sha256.der: pkcs7_data_cert_signed_sha256.der
...
+					server5.key \
+					server5.key.der \
+					server2-sha256.crt \
+					server2-sha256.crt.der \
+					server2.crt \
+					server2.crt.der \
+					server2.key \
+					server2.key.der \
+					cli2.crt \
+					cli2.crt.der \
```
Comment (lpy4105) [inline tests/data_files/Makefile:1837]: `cli2.req.sha256` is missed from `all_intermediate`, so `cli2.crt[.der]` would be re-generated each time we run `make ../src/test_certs.h`

**Review by lpy4105 — COMMENTED** [review]
LGTM, will approve after the conflict is resolved.

**Review by lpy4105 — APPROVED** [review]
LGTM

**Review by daverodgman — APPROVED** [review]

**Comment by daverodgman** [comment]
Given that all these certs are either already expired (which I assume is intentional, as they've been audited), or will not expire in the 2.28 lifetime, I don't see the need to backport.

**Pull request merged**